cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/TransformExecutor.java

```diff
@@ -17,91 +17,110 @@
 package co.cask.cdap.etl.common;
 
 import co.cask.cdap.etl.api.Destroyable;
+import co.cask.cdap.etl.api.InvalidEntry;
 import co.cask.cdap.etl.api.Transformation;
-import com.google.common.collect.Lists;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import com.google.common.collect.ImmutableList;
 
-import java.util.ArrayList;
-import java.util.Collection;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
+
 /**
  * Executes Transforms one iteration at a time, tracking how many records were input into and output from
  * each transform.
  *
  * @param <IN> the type of input object to the first transform
- * @param <OUT> the type of object output by the last transform
+ *
  */
-public class TransformExecutor<IN, OUT> implements Destroyable {
-
-
-  private static final Logger LOG = LoggerFactory.getLogger(TransformExecutor.class);
-
-  private final List<TransformDetail> transforms;
-  private final List<DefaultEmitter> emitters;
-
-
-  public TransformExecutor(List<TransformDetail> transformDetailList) {
-    int numTransforms = transformDetailList.size();
-    this.transforms = new ArrayList<>(numTransforms);
-    this.emitters = Lists.newArrayListWithCapacity(numTransforms);
-
-    for (TransformDetail transformDetail : transformDetailList) {
-      this.transforms.add(new TransformDetail(transformDetail,
-                                              new TrackedTransform(transformDetail.getTransformation(),
-                                                                   transformDetail.getMetrics())));
-      this.emitters.add(new DefaultEmitter(transformDetail.getMetrics()));
+public class TransformExecutor<IN> implements Destroyable {
+
+  private final String start;
+  private final Map<String, List<String>> connectionsMap;
+  private final TransformDetail trackedTransformDetail;
+  private final DefaultEmitter<Object> defaultEmitter;
```

Comment:
should not be re-using the same emitter for every stage.  Each stage needs its own emitter.